### PR TITLE
[FrameworkBundle] Move mailer configuration to php

### LINKED ISSUE #37186: Move internal Container configuration from XML to PHP
In Symfony 6, I we will promote usage of configuration written in PHP instead of YAML. For third-party bundles and core, we should do the same, replacing XML with PHP. Doing so would remove the need for the XML lib for core.

The biggest advantage is auto-completion with any modern IDE without explicit support for Symfony, and probably one less thing to learn (how to configure things in YAML/XML).

Be warned that semantic configuration will stay in YAML for now (as using PHP is much harder there).

#36778 did the work for the Twig bundle as an example.

We now need help from the community to move all the other bundles:

**PLEASE, comment here BEFORE starting to work on something and WAIT for the confirmation to avoid duplicate work.**

- [x] src/Symfony/Bundle/DebugBundle/Resources/config/services.xml (@jschaedl)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/annotations.xml (@magnetik)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/assets.xml (@NguyenTruongLinh)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/cache*.xml (@iamvar)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/collectors.xml (@hvt)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/console.xml (@AhmedRaafat14)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/debug*.xml (@TuxBoy)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml (@benji07)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/esi.xml (@tuanminhgp)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/form*.xml (@misekai)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/fragment*.xml (@idetox)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/http*.xml (@IonBazan)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/identity_translator.xml (@smmd)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/lock.xml (@tomasjav)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer*.xml (@instabledesign)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger*.xml (@hyoa)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/mime_type.xml (@GromNaN)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier*.xml (@szepczynski)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/profiling.xml (@hvt)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/property_access.xml (@qneyrat)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/property_info.xml (@qneyrat)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/request.xml (@dangkhoagms)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/routing*.xml (@phamuyentri)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/secrets.xml (@GromNaN)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/security_csrf.xml (@Noweh)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.xml (@mamontovdmitriy)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml (@rvanlaak)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/session.xml (@cocorambo)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/ssi.xml (@50bhan)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/test.xml (@GaryPEGEOT)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/translation*.xml (@malteschlueter)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/validator*.xml (@simivar)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/web*.xml (@ck-developer)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/workflow.xml (@lyrixx)
- [x] src/Symfony/Bundle/SecurityBundle/Resources/config/collectors.xml (@JudicaelR)
- [x] src/Symfony/Bundle/SecurityBundle/Resources/config/console.xml (@JudicaelR)
- [x] src/Symfony/Bundle/SecurityBundle/Resources/config/guard.xml (@JudicaelR)
- [x] src/Symfony/Bundle/SecurityBundle/Resources/config/security*.xml (@qneyrat)
- [x] src/Symfony/Bundle/SecurityBundle/Resources/config/templating_twig.xml (@ck-developer)
- [x] src/Symfony/Bundle/WebProfilerBundle/Resources/config/*.xml (@jschaedl)

### PR DESCRIPTION
**Pull request by instabledesign**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      |no
| New feature?  | no 
| Deprecations? | no 
| Tickets       | Part of #37186
| License       | MIT

Move mailer configuration file to php

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer_transports.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            ->parent('mailer.transport_factory.abstract')
+            ->tag('mailer.transport_factory')
+
+        ->set('mailer.transport_factory.sendmail', SendmailTransportFactory::class)
+            ->parent('mailer.transport_factory.abstract')
+            ->tag('mailer.transport_factory')
+
+        ->set('mailer.transport_factory.smtp', EsmtpTransportFactory::class)
+            ->parent('mailer.transport_factory.abstract')
+            ->tag('mailer.transport_factory', ['priority' => -100])
```
Comment (instabledesign) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer_transports.php:69]: should i set string or int value for the priority attribute?

Comment (fabpot) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer_transports.php:69]: int is fine

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            ->args([
+                service('mailer.transports'),
+                abstract_arg('message bus'),
+                service('event_dispatcher')->ignoreOnInvalid(),
+            ])
+        ->alias('mailer', 'mailer.mailer')
+        ->alias(MailerInterface::class, 'mailer.mailer')
+
+        ->set('mailer.transports', Transports::class)
+            ->factory([service('mailer.transport_factory'), 'fromStrings'])
```
Comment (instabledesign) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer.php:36]: reviewer: can you double check this syntax, thanks

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            ->args([
+                [], //transports
+            ])
+
+        ->set('mailer.transport_factory', Transport::class)
+            ->args([tagged_iterator('mailer.transport_factory')])
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer.php:42]: please use multi-line args consistently everywhere like

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer_debug.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+return static function (ContainerConfigurator $container) {
+    $container->services()
+        ->set('mailer.data_collector', MessageDataCollector::class)
+            ->args([
+                service('mailer.logger_message_listener'),
+            ])
+        ->tag('data_collector', [
+            'id' => 'mailer',
+            'template' => '@WebProfiler/Collector/mailer.html.twig',
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer_debug.php:24]: please change the order of template and id to be consistent with the other data_collector tags

**Review by Tobion — CHANGES_REQUESTED** [review]

**Comment by instabledesign** [comment]
Fixup done 
tests failed look unrelated

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                abstract_arg('message bus'),
+                service('event_dispatcher')->ignoreOnInvalid(),
+            ])
+        ->alias('mailer', 'mailer.mailer')
+        ->alias(MailerInterface::class, 'mailer.mailer')
+
+        ->set('mailer.transports', Transports::class)
+            ->factory([service('mailer.transport_factory'), 'fromStrings'])
+            ->args([
+                [], //transports
```
[the remark this answers is not in the document]
Comment (instabledesign) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer.php]: ✅

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer.php (hunk 3 above)
Comment (instabledesign) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer.php]: ✅

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer_debug.php (hunk 4 above)
Comment (instabledesign) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer_debug.php:23]: ✅

**Review by Tobion — APPROVED** [review]

**Comment by Tobion** [comment]
Thank you @instabledesign.

**Pull request merged**